File: mobile/ios/BroadcastExtension/LiveScanStore.swift
import Foundation

/// Cross-process store shared by the main app and the ReplayKit broadcast
/// extension via App Group UserDefaults. The extension writes finds; the app
/// reads them and merges into the catalog.
enum LiveScanStore {
  static let appGroupId = "group.com.streamsnap.ai"
  static let darwinName = "com.streamsnap.ai.liveScan.updated" as CFString

  private enum Key {
    static let broadcasting = "ss.live.broadcasting"
    static let startedAt = "ss.live.startedAt"
    static let lastFrameAt = "ss.live.lastFrameAt"
    static let scanCount = "ss.live.scanCount"
    static let findCount = "ss.live.findCount"
    static let okCount = "ss.live.okCount"
    static let failCount = "ss.live.failCount"
    static let skipCount = "ss.live.skipCount"
    static let lastError = "ss.live.lastError"
    static let lastPhase = "ss.live.lastPhase"
    static let lastStatus = "ss.live.lastStatus"
    static let lastBody = "ss.live.lastBody"
    static let lastJpegBytes = "ss.live.lastJpegBytes"
    static let products = "ss.live.products"
    static let sessionToken = "ss.live.sessionToken"
    static let installId = "ss.live.installId"
    static let workerUrl = "ss.live.workerUrl"
  }

  static var defaults: UserDefaults? {
    UserDefaults(suiteName: appGroupId)
  }

  static func beginSession() {
    guard let defaults else { return }
    defaults.set(true, forKey: Key.broadcasting)
    defaults.set(Date().timeIntervalSince1970, forKey: Key.startedAt)
    defaults.set(0, forKey: Key.scanCount)
    defaults.set(0, forKey: Key.findCount)
    defaults.set(0, forKey: Key.okCount)
    defaults.set(0, forKey: Key.failCount)
    defaults.set(0, forKey: Key.skipCount)
    defaults.set("started", forKey: Key.lastPhase)
    defaults.set(0, forKey: Key.lastStatus)
    defaults.set(0, forKey: Key.lastJpegBytes)
    defaults.removeObject(forKey: Key.lastError)
    defaults.removeObject(forKey: Key.lastBody)
    defaults.removeObject(forKey: Key.lastFrameAt)
    defaults.set(Data("[]".utf8), forKey: Key.products)
    defaults.synchronize()
    ping()
  }

  static func endSession() {
    defaults?.set(false, forKey: Key.broadcasting)
    defaults?.set("stopped", forKey: Key.lastPhase)
    defaults?.synchronize()
    ping()
  }

  static func writeCredentials(token: String?, installId: String, workerUrl: String) {
    guard let defaults else { return }
    if let token, !token.isEmpty {
      defaults.set(token, forKey: Key.sessionToken)
    } else {
      defaults.removeObject(forKey: Key.sessionToken)
    }
    defaults.set(installId, forKey: Key.installId)
    defaults.set(workerUrl, forKey: Key.workerUrl)
    defaults.synchronize()
  }

  static func credentials() -> (token: String?, installId: String, workerUrl: String) {
    let defaults = defaults
    return (
      defaults?.string(forKey: Key.sessionToken),
      defaults?.string(forKey: Key.installId) ?? "mob-unknown",
      defaults?.string(forKey: Key.workerUrl) ?? "https://streamsnap-lens.na0ryank0.workers.dev"
    )
  }

  static func recordScan(error: String? = nil) {
    recordEvent(phase: error == nil ? "ok" : "fail", error: error, incrementScan: true, incrementOk: error == nil, incrementFail: error != nil)
  }

  static func recordEvent(
    phase: String,
    error: String? = nil,
    status: Int? = nil,
    body: String? = nil,
    jpegBytes: Int? = nil,
    incrementScan: Bool = false,
    incrementSkip: Bool = false,
    incrementOk: Bool = false,
    incrementFail: Bool = false
  ) {
    guard let defaults else { return }
    if incrementScan {
      defaults.set(defaults.integer(forKey: Key.scanCount) + 1, forKey: Key.scanCount)
      defaults.set(Date().timeIntervalSince1970, forKey: Key.lastFrameAt)
    }
    if incrementSkip { defaults.set(defaults.integer(forKey: Key.skipCount) + 1, forKey: Key.skipCount) }
    if incrementOk { defaults.set(defaults.integer(forKey: Key.okCount) + 1, forKey: Key.okCount) }
    if incrementFail { defaults.set(defaults.integer(forKey: Key.failCount) + 1, forKey: Key.failCount) }
    defaults.set(phase, forKey: Key.lastPhase)
    if let status { defaults.set(status, forKey: Key.lastStatus) }
    if let jpegBytes { defaults.set(jpegBytes, forKey: Key.lastJpegBytes) }
    if let body {
      defaults.set(String(body.prefix(500)), forKey: Key.lastBody)
    }
    if let error {
      defaults.set(error, forKey: Key.lastError)
    } else if incrementOk {
      defaults.removeObject(forKey: Key.lastError)
    }
    defaults.synchronize()
    ping()
  }

  @discardableResult
  static func upsertProducts(_ incoming: [[String: Any]]) -> Int {
    guard let defaults else { return 0 }
    var existing = products()
    let now = Date().timeIntervalSince1970 * 1000
    var added = 0

    for raw in incoming {
      guard let title = raw["title"] as? String, !title.isEmpty else { continue }
      let asin = raw["asin"] as? String
      let url = raw["url"] as? String ?? ""
      let key = (asin?.isEmpty == false ? asin! : url)
      guard !key.isEmpty else { continue }

      if let idx = existing.firstIndex(where: { item in
        let existingKey = (item["asin"] as? String).flatMap { $0.isEmpty ? nil : $0 }
          ?? (item["url"] as? String ?? "")
        return existingKey == key
      }) {
        var item = existing[idx]
        item["seenCount"] = (item["seenCount"] as? Int ?? 1) + 1
        item["lastSeenAt"] = now
        if let price = raw["price"] as? String { item["price"] = price }
        if let imageUrl = raw["imageUrl"] as? String { item["imageUrl"] = imageUrl }
        existing[idx] = item
      } else {
        var item = raw
        item["id"] = item["id"] as? String ?? "live-\(Int(now))-\(added)"
        item["seenCount"] = 1
        item["firstSeenAt"] = now
        item["lastSeenAt"] = now
        existing.insert(item, at: 0)
        added += 1
      }
    }

    if existing.count > 80 {
      existing = Array(existing.prefix(80))
    }

    if let data = try? JSONSerialization.data(withJSONObject: existing) {
      defaults.set(data, forKey: Key.products)
    }
    defaults.set(existing.count, forKey: Key.findCount)
    defaults.synchronize()
    ping()
    return added
  }

  static func products() -> [[String: Any]] {
    guard let data = defaults?.data(forKey: Key.products),
          let json = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
    else { return [] }
    return json
  }

  static func snapshot(screenCaptured: Bool) -> [String: Any] {
    let defaults = defaults
    let creds = credentials()
    return [
      "available": defaults != nil,
      "broadcasting": defaults?.bool(forKey: Key.broadcasting) ?? false,
      "screenCaptured": screenCaptured,
      "scanCount": defaults?.integer(forKey: Key.scanCount) ?? 0,
      "findCount": defaults?.integer(forKey: Key.findCount) ?? 0,
      "okCount": defaults?.integer(forKey: Key.okCount) ?? 0,
      "failCount": defaults?.integer(forKey: Key.failCount) ?? 0,
      "skipCount": defaults?.integer(forKey: Key.skipCount) ?? 0,
      "lastError": defaults?.string(forKey: Key.lastError) as Any,
      "lastPhase": defaults?.string(forKey: Key.lastPhase) as Any,
      "lastStatus": defaults?.integer(forKey: Key.lastStatus) ?? 0,
      "lastBody": defaults?.string(forKey: Key.lastBody) as Any,
      "lastJpegBytes": defaults?.integer(forKey: Key.lastJpegBytes) ?? 0,
      "startedAt": defaults?.object(forKey: Key.startedAt) as Any,
      "lastFrameAt": defaults?.object(forKey: Key.lastFrameAt) as Any,
      "workerUrl": creds.workerUrl,
      "installId": creds.installId,
      "hasToken": creds.token != nil,
      "products": products()
    ]
  }

  static func ping() {
    CFNotificationCenterPostNotification(
      CFNotificationCenterGetDarwinNotifyCenter(),
      CFNotificationName(darwinName),
      nil,
      nil,
      true
    )
  }
}

File: mobile/modules/live-scan/android/src/main/java/expo/modules/livescan/LiveScanStore.kt
package expo.modules.livescan

import android.content.Context
import android.content.Intent
import org.json.JSONArray
import org.json.JSONObject

/**
 * SharedPreferences store for live-scan session state and finds.
 * The foreground service writes; the Expo module reads and emits onUpdate.
 */
internal object LiveScanStore {
  const val ACTION_UPDATED = "com.streamsnap.ai.LIVE_SCAN_UPDATED"
  private const val PREFS = "ss_live_scan"

  private object Key {
    const val BROADCASTING = "ss.live.broadcasting"
    const val STARTED_AT = "ss.live.startedAt"
    const val LAST_FRAME_AT = "ss.live.lastFrameAt"
    const val SCAN_COUNT = "ss.live.scanCount"
    const val FIND_COUNT = "ss.live.findCount"
    const val OK_COUNT = "ss.live.okCount"
    const val FAIL_COUNT = "ss.live.failCount"
    const val SKIP_COUNT = "ss.live.skipCount"
    const val LAST_ERROR = "ss.live.lastError"
    const val LAST_PHASE = "ss.live.lastPhase"
    const val LAST_STATUS = "ss.live.lastStatus"
    const val LAST_BODY = "ss.live.lastBody"
    const val LAST_JPEG_BYTES = "ss.live.lastJpegBytes"
    const val PRODUCTS = "ss.live.products"
    const val SESSION_TOKEN = "ss.live.sessionToken"
    const val INSTALL_ID = "ss.live.installId"
    const val WORKER_URL = "ss.live.workerUrl"
  }

  private fun prefs(context: Context) =
    context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

  fun beginSession(context: Context) {
    prefs(context).edit()
      .putBoolean(Key.BROADCASTING, true)
      .putLong(Key.STARTED_AT, System.currentTimeMillis())
      .putInt(Key.SCAN_COUNT, 0)
      .putInt(Key.FIND_COUNT, 0)
      .putInt(Key.OK_COUNT, 0)
      .putInt(Key.FAIL_COUNT, 0)
      .putInt(Key.SKIP_COUNT, 0)
      .putString(Key.LAST_PHASE, "started")
      .putInt(Key.LAST_STATUS, 0)
      .putInt(Key.LAST_JPEG_BYTES, 0)
      .remove(Key.LAST_ERROR)
      .remove(Key.LAST_BODY)
      .remove(Key.LAST_FRAME_AT)
      .putString(Key.PRODUCTS, "[]")
      .apply()
    ping(context)
  }

  fun endSession(context: Context) {
    prefs(context).edit()
      .putBoolean(Key.BROADCASTING, false)
      .putString(Key.LAST_PHASE, "stopped")
      .apply()
    ping(context)
  }

  fun writeCredentials(context: Context, token: String?, installId: String, workerUrl: String) {
    val editor = prefs(context).edit()
    if (!token.isNullOrEmpty()) {
      editor.putString(Key.SESSION_TOKEN, token)
    } else {
      editor.remove(Key.SESSION_TOKEN)
    }
    editor.putString(Key.INSTALL_ID, installId)
    editor.putString(Key.WORKER_URL, workerUrl)
    editor.apply()
  }

  data class Credentials(val token: String?, val installId: String, val workerUrl: String)

  fun credentials(context: Context): Credentials {
    val p = prefs(context)
    return Credentials(
      token = p.getString(Key.SESSION_TOKEN, null),
      installId = p.getString(Key.INSTALL_ID, null) ?: "mob-unknown",
      workerUrl = p.getString(Key.WORKER_URL, null)
        ?: "https://streamsnap-lens.na0ryank0.workers.dev"
    )
  }

  fun recordEvent(
    context: Context,
    phase: String,
    error: String? = null,
    status: Int? = null,
    body: String? = null,
    jpegBytes: Int? = null,
    incrementScan: Boolean = false,
    incrementSkip: Boolean = false,
    incrementOk: Boolean = false,
    incrementFail: Boolean = false
  ) {
    val p = prefs(context)
    val editor = p.edit()
    if (incrementScan) {
      editor.putInt(Key.SCAN_COUNT, p.getInt(Key.SCAN_COUNT, 0) + 1)
      editor.putLong(Key.LAST_FRAME_AT, System.currentTimeMillis())
    }
    if (incrementSkip) editor.putInt(Key.SKIP_COUNT, p.getInt(Key.SKIP_COUNT, 0) + 1)
    if (incrementOk) editor.putInt(Key.OK_COUNT, p.getInt(Key.OK_COUNT, 0) + 1)
    if (incrementFail) editor.putInt(Key.FAIL_COUNT, p.getInt(Key.FAIL_COUNT, 0) + 1)
    editor.putString(Key.LAST_PHASE, phase)
    if (status != null) editor.putInt(Key.LAST_STATUS, status)
    if (jpegBytes != null) editor.putInt(Key.LAST_JPEG_BYTES, jpegBytes)
    if (body != null) editor.putString(Key.LAST_BODY, body.take(500))
    if (error != null) {
      editor.putString(Key.LAST_ERROR, error)
    } else if (incrementOk) {
      editor.remove(Key.LAST_ERROR)
    }
    val nextScans = p.getInt(Key.SCAN_COUNT, 0) + if (incrementScan) 1 else 0
    val finds = p.getInt(Key.FIND_COUNT, 0)
    val broadcasting = p.getBoolean(Key.BROADCASTING, false)
    editor.apply()
    ping(context)
    if (broadcasting && (incrementScan || incrementSkip || error != null)) {
      LiveScanNotifier.updateRunning(
        context,
        scans = nextScans,
        finds = finds,
        error = if (incrementOk) null else (error ?: p.getString(Key.LAST_ERROR, null))
      )
    }
  }

  fun upsertProducts(context: Context, incoming: List<JSONObject>): Int {
    val existing = productsJson(context)
    val now = System.currentTimeMillis().toDouble()
    var added = 0

    for (raw in incoming) {
      val title = raw.optString("title").trim()
      if (title.isEmpty()) continue
      val cleanTitle = title.lowercase()
      val asin = raw.optString("asin").trim().lowercase().ifEmpty { null }

      val alreadyExists = (0 until existing.length()).any { i ->
        val item = existing.optJSONObject(i) ?: return@any false
        val existingAsin = item.optString("asin").trim().lowercase().ifEmpty { null }
        val existingTitle = item.optString("title").trim().lowercase()
        if (asin != null && existingAsin != null) {
          asin == existingAsin
        } else {
          existingTitle == cleanTitle ||
            (cleanTitle.length > 5 && existingTitle.contains(cleanTitle)) ||
            (existingTitle.length > 5 && cleanTitle.contains(existingTitle))
        }
      }
      if (alreadyExists) continue

      if (!raw.has("id")) raw.put("id", "live-${now.toLong()}-$added")
      raw.put("firstSeenAt", now)
      raw.put("lastSeenAt", now)
      existing.put(raw)
      // Insert at front: rebuild
      added += 1
    }

    if (added == 0) return 0

    val ordered = JSONArray()
    // Newest first: last `added` items were appended; reverse the new ones to the front.
    val startNew = existing.length() - added
    for (i in existing.length() - 1 downTo startNew) {
      ordered.put(existing.getJSONObject(i))
    }
    for (i in 0 until startNew) {
      if (ordered.length() >= 80) break
      ordered.put(existing.getJSONObject(i))
    }
    while (ordered.length() > 80) {
      ordered.remove(ordered.length() - 1)
    }

    prefs(context).edit()
      .putString(Key.PRODUCTS, ordered.toString())
      .putInt(Key.FIND_COUNT, ordered.length())
      .apply()
    ping(context)
    if (prefs(context).getBoolean(Key.BROADCASTING, false)) {
      LiveScanNotifier.updateRunning(
        context,
        scans = prefs(context).getInt(Key.SCAN_COUNT, 0),
        finds = ordered.length(),
        error = null
      )
    }
    return added
  }

  fun productsJson(context: Context): JSONArray {
    return try {
      JSONArray(prefs(context).getString(Key.PRODUCTS, "[]") ?: "[]")
    } catch (_: Exception) {
      JSONArray()
    }
  }

  fun products(context: Context): List<Map<String, Any?>> {
    val arr = productsJson(context)
    return (0 until arr.length()).mapNotNull { i ->
      arr.optJSONObject(i)?.toMap()
    }
  }

  fun snapshot(context: Context, screenCaptured: Boolean): Map<String, Any?> {
    val p = prefs(context)
    val creds = credentials(context)
    val startedAtMs = p.getLong(Key.STARTED_AT, 0L)
    val lastFrameAtMs = p.getLong(Key.LAST_FRAME_AT, 0L)
    return mapOf(
      "available" to true,
      "broadcasting" to p.getBoolean(Key.BROADCASTING, false),
      "screenCaptured" to screenCaptured,
      "scanCount" to p.getInt(Key.SCAN_COUNT, 0),
      "findCount" to p.getInt(Key.FIND_COUNT, 0),
      "okCount" to p.getInt(Key.OK_COUNT, 0),
      "failCount" to p.getInt(Key.FAIL_COUNT, 0),
      "skipCount" to p.getInt(Key.SKIP_COUNT, 0),
      "lastError" to p.getString(Key.LAST_ERROR, null),
      "lastPhase" to p.getString(Key.LAST_PHASE, null),
      "lastStatus" to p.getInt(Key.LAST_STATUS, 0),
      "lastBody" to p.getString(Key.LAST_BODY, null),
      "lastJpegBytes" to p.getInt(Key.LAST_JPEG_BYTES, 0),
      "startedAt" to if (startedAtMs > 0) startedAtMs / 1000.0 else null,
      "lastFrameAt" to if (lastFrameAtMs > 0) lastFrameAtMs / 1000.0 else null,
      "workerUrl" to creds.workerUrl,
      "installId" to creds.installId,
      "hasToken" to (creds.token != null),
      "products" to products(context)
    )
  }

  fun ping(context: Context) {
    val intent = Intent(ACTION_UPDATED).setPackage(context.packageName)
    context.sendBroadcast(intent)
  }

  private fun JSONObject.toMap(): Map<String, Any?> {
    val out = mutableMapOf<String, Any?>()
    val keys = keys()
    while (keys.hasNext()) {
      val key = keys.next()
      out[key] = when (val v = opt(key)) {
        JSONObject.NULL, null -> null
        is JSONObject -> v.toMap()
        is JSONArray -> (0 until v.length()).map { v.opt(it) }
        else -> v
      }
    }
    return out
  }
}
